File: proj/v5_3out/onnx/main.py
import cv2
import numpy as np
import onnxruntime


def _sigmoid(x):
    return 1. / (1. + np.exp(-x))



class YOLOV5:

    def __init__(self, modelpath, conf_thres=0.7, nms_thres=0.8):

        self.confThreshold = conf_thres
        self.nmsThreshold = nms_thres

        # self.class_names = [x.strip() for x in open("coco.names", "r").readlines()]
        self.class_names = ["fire"]
        self.anchors = {
            8: [[1.25000, 1.62500], [2.00000, 3.75000], [4.12500, 2.87500]],
            16: [[1.87500, 3.81250], [3.87500, 2.81250], [3.68750, 7.43750]],
            32: [[3.62500, 2.81250], [4.87500, 6.18750], [11.65625, 10.18750]]
            }

        
        self.inpWidth = 640
        self.inpHeight = 640

        self.sess = onnxruntime.InferenceSession(modelpath, providers=["CUDAExecutionProvider", "CPUExecutionProvider"])

        self.innames = [x.name for x in self.sess.get_inputs()]
        self.ounames = [x.name for x in self.sess.get_outputs()]


    def drawPred(self, image, box, classid, conf):
        x,y,w,h = box

        cv2.rectangle(image, (x, y), (x+w, y+h), (255,0,0), 2)

        label = self.class_names[classid] + ":" + str(round(conf * 100)) + "%"

        cv2.putText(image, label, (x,y-10), cv2.FONT_HERSHEY_SIMPLEX, 1, (0,255,0), 2)

    def resize_img(self, image):
        neww, newh, padw, padh = self.inpWidth, self.inpHeight, 0, 0

        srch, srcw = image.shape[:2]

        if srch != srcw:
            hw_scale = srch / srcw
            if hw_scale > 1.0:
                neww = int(self.inpWidth / hw_scale)
                timg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)
                padw = int((self.inpWidth - neww) * 0.5)
                timg = cv2.copyMakeBorder(timg, 0, 0, padw, self.inpWidth-neww-padw, cv2.BORDER_CONSTANT, (114,114,114))
            else:
                newh = int(self.inpHeight * hw_scale)
                timg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)
                padh = int((self.inpHeight - newh) * 0.5)
                timg = cv2.copyMakeBorder(timg, padh, self.inpHeight-newh-padh, 0, 0, cv2.BORDER_CONSTANT, (114,114,114))
        else:
            timg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)

        return timg, neww, newh, padw, padh
    
    def make_grid(self, nx, ny, stride, anchor):
        stride = np.array(stride)
        anchor = np.array(anchor)
        xv, yv = np.meshgrid(np.arange(nx), np.arange(ny))
        grid = np.stack((xv, yv), -1)
        anchor_grid = (anchor * stride).reshape(1, len(anchor), 1, 1, 2)
        return grid, anchor_grid

    
    
    
    def postprocess(self, outs, padw, padh):
        z = []
        for out in outs:
            _, _, ny, nx, _ = out.shape
            stride = self.inpHeight / ny
            assert (stride == self.inpWidth / nx)
            anchor = self.anchors[stride]
            grid, anchor_grid = self.make_grid(nx, ny, stride, anchor)
            
            y = _sigmoid(out)
            y[..., 0:2] = (y[..., 0:2] * 2 - 0.5 + grid) * stride  # xy
            y[..., 0:1] = y[..., 0:1] - padw  # x
            y[..., 1:2] = y[..., 1:2] - padh  # y
            y[..., 2:4] = (y[..., 2:4] * 2)**2 * anchor_grid  # wh
            z.append(y.reshape(-1, 6))
        preds = np.concatenate(z, axis=0)
        return preds
            
            
    def detect(self, image):

        timg, neww, newh, padw, padh = self.resize_img(image)

        blob = cv2.dnn.blobFromImage(timg, 1.0 / 255.0, swapRB=True)
        
        outs = self.sess.run(self.ounames, {self.innames[0]:blob}) 
        
        preds = self.postprocess(outs, padw, padh)

        ratioh, ratiow = image.shape[0] / newh, image.shape[1] / neww
        boxes, ids, confs = [], [], []

        for pred in preds:
            
            if pred[4] > self.confThreshold:
                maxss = np.max(pred[5:])
                idp = np.argmax(pred[5:])
                maxss *= pred[4]
                if maxss >= self.confThreshold:
                    cx = pred[0] * ratiow
                    cy = pred[1] * ratioh
                    w = pred[2] * ratiow
                    h = pred[3] * ratioh

                    left = int(cx - 0.5 * w)
                    top = int(cy - 0.5 * h)

                    boxes.append([left, top, int(w), int(h)])
                    confs.append(maxss)
                    ids.append(idp)

        indices = cv2.dnn.NMSBoxes(boxes, confs, self.confThreshold, self.nmsThreshold)

        for idx in indices:
            self.drawPred(image, boxes[idx], ids[idx], confs[idx])
        return image

if __name__ == "__main__":

    net = YOLOV5("models/lxfire_base.onnx")

    srcimg = cv2.imread("imgs/00007.jpg")

    tarimg = net.detect(srcimg)

    cv2.imwrite("imgs/00007_out3_onnx_py.jpg", tarimg)

File: proj/v5_3out/onnx/main.cpp
#include <iostream>
#include <fstream>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/dnn.hpp>
#include <onnxruntime_cxx_api.h>

static const float anchors[3][6] = {{10.0, 13.0, 16.0, 30.0, 33.0, 23.0}, {30.0, 61.0, 62.0, 45.0, 59.0, 119.0},{116.0, 90.0, 156.0, 198.0, 373.0, 326.0}};
static const float stride[3] = { 8.0, 16.0, 32.0 };


static inline float sigmoid(float x) { return 1.0 / (1.0 + expf(-x)); }

class YOLOV5
{
public:
	YOLOV5(float confThreshold, float nmsThreshold, std::string modelpath);
	void drawPred(cv::Mat& frame, cv::Rect box, int classid, float conf);
	cv::Mat resize_img(cv::Mat img, int *neww, int *newh, int *padw, int *padh);
	void normalize_(cv::Mat img);
	void detect(cv::Mat& frame);
private:

	int inpWidth;
	int inpHeight;
	int nout;

	float confThreshold;
	float nmsThreshold;


	std::vector<float> input_image_;
	std::vector<std::string> class_names;


	std::vector<const char*> input_names{"images"};
	std::vector<const char*> output_names{"output", "364", "381"};

	Ort::Env env = Ort::Env(ORT_LOGGING_LEVEL_ERROR, "YOLOV5");

	Ort::Session *sess = nullptr;
	Ort::SessionOptions sessops = Ort::SessionOptions();

};

YOLOV5::YOLOV5(float confThreshold, float nmsThreshold, std::string modelpath)
{
	this->confThreshold = confThreshold;
	this->nmsThreshold = nmsThreshold;

	this->inpWidth = 640;
	this->inpHeight = 640;
	this->nout = 6;

	this->class_names = {"fire"};


	OrtSessionOptionsAppendExecutionProvider_CUDA(this->sessops, 0);

	this->sessops.SetGraphOptimizationLevel(ORT_ENABLE_BASIC);

	this->sess = new Ort::Session(this->env, modelpath.c_str(), this->sessops);

}

void YOLOV5::drawPred(cv::Mat& frame, cv::Rect box, int classid, float conf)
{
	cv::rectangle(frame, box, cv::Scalar(0,144,255), 2);

	std::string label = cv::format("%.2f", conf);

	label = this->class_names[classid] + ":" + label;

	cv::putText(frame, label, cv::Point(box.x,box.y-10), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(144,0,255), 2);


}

cv::Mat YOLOV5::resize_img(cv::Mat img, int *neww, int *newh, int *padw, int *padh)
{
	*neww = this->inpWidth, *newh = this->inpHeight;
	int srch = img.rows, srcw = img.cols;

	cv::Mat timg;

	if(srch != srcw)
	{
		float hw_scale = (float)srch / srcw;
		
		if(hw_scale > 1.0)
		{
			*neww = int(this->inpWidth / hw_scale);
			cv::resize(img, timg, cv::Size(*neww, *newh), cv::INTER_AREA);
			*padw = int((this->inpWidth - *neww) * 0.5);
			cv::copyMakeBorder(timg, timg, 0, 0, *padw, this->inpWidth-*neww-*padw, cv::BORDER_CONSTANT, 114);
		
		}
		else
		{
			*newh = int(this->inpHeight * hw_scale);
			cv::resize(img, timg, cv::Size(*neww, *newh), cv::INTER_AREA);
			*padh = int((this->inpHeight - *newh) * 0.5);
			cv::copyMakeBorder(timg, timg, *padh, this->inpHeight-*newh-*padh, 0, 0, cv::BORDER_CONSTANT, 114);
		
		}
		
	}
	else
	{
		cv::resize(img, timg, cv::Size(*neww, *newh), cv::INTER_AREA);
	}
	return timg;
}



void YOLOV5::normalize_(cv::Mat img)
{
	int srch = img.rows, srcw = img.cols;
	this->input_image_.resize(srcw * srch * img.channels());

	for(int c = 0; c < 3; ++c)
	{
		for(int i = 0; i < srch; ++i)
		{
			for(int j = 0; j < srcw; ++j)
			{
				float pix = img.ptr<uchar>(i)[j * 3 + 2 - c];
				this->input_image_[c*srcw*srch+i*srcw+j] = pix / 255.0;
			
			}
		}
	
	}

}


void YOLOV5::detect(cv::Mat& frame)
{
	int neww = 0, newh = 0, padw = 0, padh = 0;

	cv::Mat timg = this->resize_img(frame, &neww, &newh, &padw, &padh);


	this->normalize_(timg);

	std::array<int64_t, 4> input_shape_{1, 3, this->inpHeight, this->inpWidth};

	auto allocator_info = Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeCPU);
	Ort::Value input_tensor = Ort::Value::CreateTensor<float>(allocator_info, input_image_.data(), input_image_.size(), input_shape_.data(), input_shape_.size());

	std::vector<Ort::Value> output_tensors = this->sess->Run(Ort::RunOptions{nullptr}, &input_names[0], &input_tensor, 1, output_names.data(), output_names.size());

	std::vector<int> ids;
	std::vector<float> confs;
	std::vector<cv::Rect> boxes;

	float ratiow = (float)frame.cols / neww;
	float ratioh = (float)frame.rows / newh;

	int n = 0, q = 0, i = 0, j = 0, c = 0;
	
	for(n = 0; n < 3; ++n)
	{
		int num_grid_x = (int)(this->inpWidth / stride[n]);
		int num_grid_y = (int)(this->inpHeight / stride[n]);
		int area = num_grid_x * num_grid_y;

		for(q = 0; q < 3; ++q)
		{
			const float anchor_w = anchors[n][q * 2];
			const float anchor_h = anchors[n][q * 2 + 1];

			float *pdata = output_tensors[n].GetTensorMutableData<float>() + q * this->nout * area;
				
			for(i = 0; i < num_grid_y; ++i)
			{	
				for(j = 0; j < num_grid_x; ++j)
				{
							
					float conf = sigmoid(pdata[i * num_grid_x * this->nout + j * this->nout + 4]);
					if(conf > this->confThreshold)
					{
						float maxss = 0.0;
						int idp = 0;
						for(c = 0; c < this->class_names.size(); ++c)
						{
							float cla_s = sigmoid(pdata[i * num_grid_x * this->nout + j * this->nout + c + 5]);
							if(cla_s > maxss)
							{
								maxss = cla_s;
								idp = c;
							}
						
						}
						maxss *= conf;	
						if(maxss >= this->confThreshold)
						{
							float scx = (sigmoid(pdata[i * num_grid_x * this->nout + j * this->nout]) * 2.0 - 0.5 + j) * stride[n];
							float scy = (sigmoid(pdata[i * num_grid_x * this->nout + j * this->nout + 1]) * 2.0 - 0.5 + i) * stride[n];
							float sw = powf(sigmoid(pdata[i * num_grid_x * this->nout + j * this->nout + 2]) * 2.0, 2.0) * anchor_w;
							float sh = powf(sigmoid(pdata[i * num_grid_x * this->nout + j * this->nout + 3]) * 2.0, 2.0) * anchor_h;
							
							float cx = (scx - padw) * ratiow;
							float cy = (scy - padh) * ratioh;
							float w = sw * ratiow;
							float h = sh * ratioh;

							int left = int(cx - 0.5 * w);
							int top = int(cy - 0.5 * h);


							boxes.push_back(cv::Rect(left, top, int(w), int(h)));
							confs.push_back(maxss);
							ids.push_back(idp);
						}
					}
					
						
				}
					
			}
				


		}

	
	}
	std::vector<int> indices;
	cv::dnn::NMSBoxes(boxes, confs, this->confThreshold, this->nmsThreshold, indices);
	for(int idx : indices) this->drawPred(frame, boxes[idx], ids[idx], confs[idx]);
	


}


int main()
{
	YOLOV5 net(0.7, 0.8, "models/lxfire_base.onnx");

	cv::Mat srcimg = cv::imread("imgs/00007.jpg");

	net.detect(srcimg);

	cv::imwrite("imgs/0007_3out_onnx_cpp.jpg", srcimg);

	return 0;
}










